File: mpl_extract.c
/* GPL 2.1 © Elte 2011 */

#include <stdio.h>
#include <string.h>
#include <unistd.h>
#include <errno.h>
#include <dirent.h>

#define VERSION         "0.1"
#define SIGNATUR        "MPLS0100"

typedef struct {
    unsigned char decade, month, day, hour, minute, second;
} MTS_date;

int main (int argc, char *argv[])
{
    struct dirent  *dp = NULL;
    DIR       *dir_ptr = NULL;
    FILE *mpl_file_ptr = NULL;

    int ret = 0;

    unsigned char signatur[] = SIGNATUR, buffer[sizeof(SIGNATUR)];

    if (argc != 2) {
        fprintf(stderr, "Version: %s\n", VERSION);
        fprintf(stderr, "Usage: %s <path to folder containing *.MPL files\n", argv[0]);
        goto ERROR;
    }

    if (chdir(argv[1]) == -1) {
    fprintf(stderr, "Cannot change directory: %s %s\n", argv[1], strerror(errno));
    goto ERROR;
    }

    if ((dir_ptr = opendir (".")) == NULL) {
        fprintf (stderr, "Error reading directory: %s %s\n", argv[1], strerror(errno));
        goto ERROR;
    }

//    printf (" STREAM       DATE       TIME\n");
//    printf ("---------  ----------  --------\n");

    while ((dp = readdir(dir_ptr)) != NULL) {

        if (!strncmp ((dp->d_name) + (strlen (dp->d_name) - 4), ".MPL", 4)) {

            unsigned char file_number[2] = {0, 0},
                                num_desc = 0;
            MTS_date                this = {0, 0, 0, 0, 0, 0};

            if ((mpl_file_ptr = fopen (dp->d_name, "r")) == NULL) {
                fprintf (stderr, "Error opening file: %s %s\n", dp->d_name, strerror(errno));
                goto NEXT_FILE;
            }

            if ((fread (buffer, sizeof(SIGNATUR), 1, mpl_file_ptr) != 1) ||
                (strncmp (buffer, SIGNATUR, sizeof(SIGNATUR)))) {
                fprintf (stderr, "Could not read file signatur. Wrong filetype?\n");
                goto NEXT_FILE;
            }

            // find out how many mts files are described
            // the 66th byte contains this number
            fseek (mpl_file_ptr, 66 - sizeof(SIGNATUR) - 1, SEEK_CUR);
            ret = fgetc(mpl_file_ptr);
            if (ret != EOF) {
                num_desc = (unsigned char) ret;
            } else {
                fprintf (stderr, "Could not read contents\n");
                goto NEXT_FILE;
            }

            // jump to the first occurance of a time stamp and print it
            // iterate till all num_desc time stamps are shown
            // trailer = 50 bytes
            // mts description = 66 bytes
            // actual info starts at 9th byte of mts description
            fseek (mpl_file_ptr, -50 - 66*num_desc - 48 +2, SEEK_END);

            while (num_desc > 0) {

                int i;
                const unsigned char time_stamp_sig[] = {1, 3, 5, 0x01, 0, 0, 0, 2};

                num_desc--;
                fseek (mpl_file_ptr, 48, SEEK_CUR);

                // scan for time stamp signatur
                for (i=0; i<8; i++) {
                    if ((ret = fgetc(mpl_file_ptr)) != time_stamp_sig[i]) {
                        fprintf (stderr, "Could not parse contents\n");
                        goto NEXT_FILE;
                    }
                }

                // scan time stamp
                if ((ret = fscanf (mpl_file_ptr, "%c%c\x1E%*1[ ]%c%c%c%c%c%c",
                              &file_number[0], &file_number[1],
                              &this.decade, &this.month, &this.day,
                              &this.hour, &this.minute, &this.second)) != 8) {
                    fprintf (stderr, "Could not parse time stamp\n");
                    goto NEXT_FILE;
                }
                printf ("%.5d.MTS  20%.2X/%.2X/%.2X  %.2X:%.2X:%.2X\n",
                        (file_number[0]<<8) + file_number[1],
                        this.decade, this.month, this.day,
                        this.hour, this.minute, this.second);
            }
NEXT_FILE:
            fclose (mpl_file_ptr);
        }
    }
ERROR:
    closedir (dir_ptr);
}

File: mpl_extract.py
#!/usr/bin/env python3

import sys
import os
import datetime
import json


class MplFileException(Exception):
    pass


class MtsInfo:
    def __init__(self, seek_pos, info):
        self.seek_pos = seek_pos
        self.__info = info
        self.dirty = False

    def __getitem__(self, key):
        return self.__info[key]

    def __setitem__(self, key, value):
        self.__info[key] = value
        self.dirty = True


class MplFile:
    SIGNATURE = b"MPLS0100"

    def __init__(self, filepath):
        self.filepath = filepath
        self.timezone_filepath = filepath + ".timezone.json"
        self.db = {}
        self.tz_dict = {}
        self.read()

    def get_mts(self, mts_name):
        return self.db[mts_name]

    def read(self):
        try:
            with open(self.timezone_filepath, "r") as f:
                self.tz_dict = json.load(f)
        except FileNotFoundError:
            pass

        with open(self.filepath, "rb") as f:

            if f.read(len(self.SIGNATURE)) != self.SIGNATURE:
                raise MplFileException("Could not read file signature. Wrong filetype?")

            # find out how many mts files are described
            # the 66th byte contains this number
            f.seek(66 - len(self.SIGNATURE) - 1, os.SEEK_CUR)
            ret = f.read(1)
            if len(ret) != 1:
                raise MplFileException("Could not read contents")
            num_desc = ord(ret)

            # jump to the first occurance of a time stamp and print it
            # iterate till all num_desc time stamps are shown
            # trailer = 50 bytes
            # mts description = 66 bytes
            # actual info starts at 9th byte of mts description
            f.seek(-50 - 66*num_desc - 36 + 2, os.SEEK_END)

            while num_desc > 0:
                time_stamp_sig = b'\x01\x03\x05\x01\x00\x00\x00\x02'

                num_desc -= 1
                f.seek(36, os.SEEK_CUR)

                # scan for time stamp signature
                assert f.read(len(time_stamp_sig)) == time_stamp_sig

                timestamp_seek_pos = f.tell()

                # scan time stamp
                mts_filename = "{:05d}.MTS".format(int.from_bytes(f.read(2), byteorder='big'))
                assert f.read(1) == b'\x1E'
                year = int("{:x}".format(int.from_bytes(f.read(2), byteorder='big')))

                datetime_values = (
                    int("{:x}".format(ord(f.read(1)))) for _ in range(5)
                )

                dt = datetime.datetime(year, *datetime_values)
                tzoffset_seconds = self.tz_dict.get(mts_filename)
                if tzoffset_seconds is not None:
                    dt = dt.replace(tzinfo=datetime.timezone(datetime.timedelta(seconds=tzoffset_seconds)))

                assert f.read(2) in {b'\x90\x0A', b'\x90\x0C'}
                assert f.read(10) == f"{dt.year:4d}.{dt.month:2d}.{dt.day:2d}".encode('ascii')

                self.db[mts_filename] = MtsInfo(seek_pos=timestamp_seek_pos, info={
                    "datetime": dt,
                })

    def write(self):
        if not any(info.dirty for info in self.db.values()):
            return

        with open(self.filepath, "r+b") as f:
            for mts_filename, info in self.db.items():
                if not info.dirty:
                    continue

                f.seek(info.seek_pos)
                dt = info["datetime"]

                assert f.read(2) == int(mts_filename[:-4]).to_bytes(2, byteorder='big')
                assert f.read(1) == b'\x1E'
                f.write(int(f"{dt.year}", 16).to_bytes(2, byteorder='big'))

                for val in (dt.month, dt.day, dt.hour, dt.minute, dt.second):
                    f.write(bytes((int(f"{val}", 16),)))

                assert f.read(2) in {b'\x90\x0A', b'\x90\x0C'}

                f.write(f"{dt.year:4d}.{dt.month:2d}.{dt.day:2d}".encode('ascii'))

                if dt.tzinfo is None:
                    self.tz_dict.pop(mts_filename, None)
                else:
                    self.tz_dict[mts_filename] = dt.tzinfo.utcoffset(None).total_seconds()

                info.dirty = False

        with open(self.timezone_filepath, "w") as f:
            json.dump(self.tz_dict, f)



class MplDirectory:
    def __init__(self, path):
        self.path = path
        self.db = {}
        self.mpl_files = {}

    def get_mts(self, mts_name):
        return self.db[mts_name]

    def print(self):
        for mts_name, info in self.db.items():
            print("{}: {} {}".format(mts_name, info['datetime'], info['datetime'].tzinfo))

    def read(self):
        for root, dirs, files in os.walk(self.path):
            for file in files:
                if file.endswith(".MPL"):
                    filepath = os.path.join(root, file)
                    self.mpl_files[filepath] = mpl = MplFile(filepath)
                    self.db.update(mpl.db)

    def write(self):
        for mpl in self.mpl_files.values():
            mpl.write()


def main():
    mpl = MplDirectory(path=sys.argv[1])
    mpl.read()

    dt = mpl.get_mts("00127.MTS")["datetime"]
    # mpl.get_mts("00127.MTS")["datetime"] = dt.replace(year=2520, month=6, day=2, hour=23, minute=50, second=51, tzinfo=datetime.timezone(datetime.timedelta(hours=3)))

    mpl.print()
    mpl.write()


if __name__ == "__main__":
    main()
